Task 2172 associate license org

vichugofsl opened the pull request:
# Description

update the method update_licensors according to the new logic to associate a licensor with a fileset. It has considered the fields called `LORA`, `TextAgreement` and `OtherAgreement`. Additionally,  It has considered the field: `Reg_Recording_Status`.

@bradflood , I have added the feature to create a new licensor (organization) if the licensor retrieved from XML does not exist in the DBP database.

# Task
[User Story 2172](https://dev.azure.com/keyholesoftware/FCBH%20-%20Bible%20Brain/_workitems/edit/2172): associate license orgs for LORA
[User Story 2174](https://dev.azure.com/keyholesoftware/FCBH%20-%20Bible%20Brain/_workitems/edit/2174): associate org licenses for "Text Only"
[User Story 2173](https://dev.azure.com/keyholesoftware/FCBH%20-%20Bible%20Brain/_workitems/edit/2173): associate org licenses with TextAgreement

# How to test
I should run the command:

```shell
python3 load/LoadOrganizations.py test
```

- outcome
```shell
Config 'test' is loaded.
Database 'dbp_2024_03_20' is opened.
migration stage: B
migration stage: [B]
LPTS Size. Prior: 13, Current: 13
LPTS Extract with 13 records and 13 BibleIds is loaded.
num filelists 15420
WARN Agreement Type is LORA. Licensor Elam Ministries has been added as the audio licensor, fileset id: AZBEMV. We cannot know if the licensor is also the text licensor. Please modify manually if necessary
WARN Agreement Type is LORA. Licensor Elam Ministries has been added as the audio licensor, fileset id: AZBEMV. We cannot know if the licensor is also the text licensor. Please modify manually if necessary
ERROR ambiguous agreement and it needs to be fixed for fileset id: BAXABCN1DA
ERROR ambiguous agreement and it needs to be fixed for fileset id: BAXABCN1DA-opus16
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Davar Partners International and co licensor: The Love Fellowship, fileset id: DGODPIN_ET
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Davar Partners International and co licensor: The Love Fellowship, fileset id: DGODPIN_ET-usx
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Davar Partners International and co licensor: The Love Fellowship, fileset id: DGODPIN1DA
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Davar Partners International and co licensor: The Love Fellowship, fileset id: DGODPIN1DA-opus16
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Davar Partners International and co licensor: The Love Fellowship, fileset id: DGODPIP2DV
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Silakkuagvik Communications, Inc. (KBRW) and co licensor: Canadian Bible Society, fileset id: ESIABSN_ET
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Silakkuagvik Communications, Inc. (KBRW) and co licensor: Canadian Bible Society, fileset id: ESIABSN_ET-json
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Silakkuagvik Communications, Inc. (KBRW) and co licensor: Canadian Bible Society, fileset id: ESIABSN_ET-usx
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Silakkuagvik Communications, Inc. (KBRW) and co licensor: Canadian Bible Society, fileset id: ESIABSN1DA
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Silakkuagvik Communications, Inc. (KBRW) and co licensor: Canadian Bible Society, fileset id: ESIABSN1DA-opus16
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Silakkuagvik Communications, Inc. (KBRW) and co licensor: Canadian Bible Society, fileset id: ESIABSN1DA16
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Davar Partners International and co licensor: The Love Fellowship, fileset id: KFYDPIN_ET
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Davar Partners International and co licensor: The Love Fellowship, fileset id: KFYDPIN_ET-usx
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Davar Partners International and co licensor: The Love Fellowship, fileset id: KFYDPIN1DA
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Davar Partners International and co licensor: The Love Fellowship, fileset id: KFYDPIN1DA-opus16
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Davar Partners International and co licensor: The Love Fellowship, fileset id: KFYDPIP2DV
ERROR ambiguous agreement and it needs to be fixed for fileset id: KNJSBI
ERROR ambiguous agreement and it needs to be fixed for fileset id: KNJSBIN_ET
ERROR ambiguous agreement and it needs to be fixed for fileset id: KNJSBIN_ET-json
ERROR ambiguous agreement and it needs to be fixed for fileset id: KNJSBIN2DA
ERROR ambiguous agreement and it needs to be fixed for fileset id: KNJSBIN2DA-opus16
ERROR ambiguous agreement and it needs to be fixed for fileset id: KNJSBIN2DA16
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Folk&Språk and Swedish Mission Alliance and co licensor: SIL, fileset id: MOXFSA
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Folk&Språk and Swedish Mission Alliance and co licensor: SIL, fileset id: MOXFSAN_ET
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Folk&Språk and Swedish Mission Alliance and co licensor: SIL, fileset id: MOXFSAN_ET-json
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Folk&Språk and Swedish Mission Alliance and co licensor: SIL, fileset id: MOXFSAN_ET-usx
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Folk&Språk and Swedish Mission Alliance and co licensor: SIL, fileset id: MOXFSAN1DA
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Folk&Språk and Swedish Mission Alliance and co licensor: SIL, fileset id: MOXFSAN1DA-opus16
WARN Agreement Type is 'Other', which is not currently being processed. licensor: Folk&Språk and Swedish Mission Alliance and co licensor: SIL, fileset id: MOXFSAN1DA16
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES ('30', 'b9bbe8f9cd88', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='b9bbe8f9cd88' AND organization_id='734' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES ('30', '662259462eb7', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='662259462eb7' AND organization_id='734' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES ('30', 'b6f9756a66b8', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='b6f9756a66b8' AND organization_id='734' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES ('30', '42143125157f', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='42143125157f' AND organization_id='734' AND organization_role='2';
INSERT IGNORE INTO organizations (slug) VALUES ('SIL');
SET @SIL = LAST_INSERT_ID();
INSERT IGNORE INTO organization_translations (organization_id, name, description, language_id) VALUES (@SIL, 'SIL', 'SIL', '6414');
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES (@SIL, 'b343589b5cae', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='b343589b5cae' AND organization_id='30' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES (@SIL, '7ea398455d03', '2');
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES ('249', '7ea398455d03', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='7ea398455d03' AND organization_id='30' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES (@SIL, 'd804f1302c97', '2');
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES ('249', 'd804f1302c97', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='d804f1302c97' AND organization_id='30' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES (@SIL, '7ab32fbfd94b', '2');
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES ('249', '7ab32fbfd94b', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='7ab32fbfd94b' AND organization_id='30' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES (@SIL, '8776df2d7664', '2');
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES ('249', '8776df2d7664', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='8776df2d7664' AND organization_id='30' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES ('9', '3bfba6cfdf73', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='3bfba6cfdf73' AND organization_id='30' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES ('9', 'c818c595d6d7', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='c818c595d6d7' AND organization_id='30' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES ('9', '918619867d45', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='918619867d45' AND organization_id='30' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES ('9', 'aeae66dfc026', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='aeae66dfc026' AND organization_id='30' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES ('9', '613aafa8f497', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='613aafa8f497' AND organization_id='30' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES ('9', '3c756729aed2', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='3c756729aed2' AND organization_id='30' AND organization_role='2';
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='427fd8643264' AND organization_id='177' AND organization_role='2';
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='eb3056cc76d6' AND organization_id='177' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES (@SIL, '726de5d29b88', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='726de5d29b88' AND organization_id='30' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES (@SIL, 'a243241bb70e', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='a243241bb70e' AND organization_id='30' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES (@SIL, 'c842c89d06a7', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='c842c89d06a7' AND organization_id='30' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES (@SIL, '9d46c4b70f17', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='9d46c4b70f17' AND organization_id='30' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES ('9', '37bf6917bd30', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='37bf6917bd30' AND organization_id='30' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES ('9', 'b85d591644e1', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='b85d591644e1' AND organization_id='30' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES ('9', 'ad57096de342', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='ad57096de342' AND organization_id='30' AND organization_role='2';
INSERT IGNORE INTO bible_fileset_copyright_organizations (organization_id, hash_id, organization_role) VALUES ('9', '695a8b4fe148', '2');
DELETE FROM bible_fileset_copyright_organizations WHERE hash_id='695a8b4fe148' AND organization_id='30' AND organization_role='2';
insert bible_fileset_copyright_organizations 27
delete bible_fileset_copyright_organizations 25
insert organizations 1
insert organization_translations 1
Transactions ./Trans-test-orgs.sql
Config 'test' is loaded.
AWSSession. assume role arn: arn:aws:iam::869054869504:role/dbp-etl-dev
Created role arn:aws:iam::869054869504:role/dbp-etl-dev based session for s3.
Enter password: 
SQLBATCH: 
SQLBATCH execution time 0.55 sec for test-orgs
```

vichugofsl commented:
@bradflood , you can see the latest changes in the second commit: https://github.com/faithcomesbyhearing/dbp-etl/pull/130/commits/d2c96cbd7d0a452e9f388636ead3467e90cae071

bradflood commented on load/LoadOrganizations.py:
```diff
@@ -98,101 +95,182 @@ def hasDamIds(self, languageRecord, typeCode):
+			return None
+
+		if language_record.Has_Other_Agreement():
+			print("WARN licensor: %s and co licensor: %s cannot be processed due to 'other Agreement' for fileset id: %s" % (language_record.Licensor(), language_record.CoLicensor(), fileset_id))
```
reword
print("WARN Agreement Type is Other, which is not currently being processed.  licensor: %s and co licensor: %s,   fileset id: %s"

vichugofsl replied:
@bradflood I have fixed it. See: https://github.com/faithcomesbyhearing/dbp-etl/pull/130/files#diff-0628aedffa821d1b9acf4c81d50c756fe72674fcc7ca34c47e0a5b0a65c60aceR105

bradflood commented on load/LoadOrganizations.py:
```diff
@@ -98,101 +95,182 @@ def hasDamIds(self, languageRecord, typeCode):
+			return licensors
+
+		if language_record.Has_Lora_Agreement() and type_code == "text":
+			self.add_licensor_if_not_empty(licensors, language_record.CoLicensor(), fileset_id, "co-licensor")
```
don't call if coLicensor is empty

vichugofsl replied:
@bradflood , It has been updated as you can see in the following link: https://github.com/faithcomesbyhearing/dbp-etl/pull/130/files#diff-0628aedffa821d1b9acf4c81d50c756fe72674fcc7ca34c47e0a5b0a65c60aceR117

bradflood commented on load/LoadOrganizations.py:
```diff
@@ -98,101 +95,182 @@ def hasDamIds(self, languageRecord, typeCode):
+		if language_record.Has_Lora_Agreement() and type_code == "text":
+			self.add_licensor_if_not_empty(licensors, language_record.CoLicensor(), fileset_id, "co-licensor")
+			if language_record.CoLicensor() == None and language_record.Licensor() != None:
+				print("WARN %s has ambiguos text agreement, licensor: %s may or may not be the text licensor" % (fileset_id, language_record.Licensor()))
```
reword
"WARN Agreement Type is LORA. Licensor %s has been added as the audio licensor.  We cannot know if the licensor is also the text licensor. Please modify manually if necessary"

vichugofsl replied:
@bradflood , It has been updated. See: https://github.com/faithcomesbyhearing/dbp-etl/pull/130/files#diff-0628aedffa821d1b9acf4c81d50c756fe72674fcc7ca34c47e0a5b0a65c60aceR121

vichugofsl commented on load/LoadOrganizations.py:
```diff
@@ -98,101 +95,182 @@ def hasDamIds(self, languageRecord, typeCode):
+	def get_lpts_licensors(self, language_record, fileset_id, type_code):
+		licensors = []
+		if language_record.Has_Ambiguous_Agreement():
+			print("WARN ambiguos agreement and it needs to be fixed for fileset id: %s" % (fileset_id))
```
@bradflood , it has been fixed. See: https://github.com/faithcomesbyhearing/dbp-etl/pull/130/files#diff-0628aedffa821d1b9acf4c81d50c756fe72674fcc7ca34c47e0a5b0a65c60aceR101

vichugofsl commented on load/LoadOrganizations.py:
```diff
@@ -98,101 +95,182 @@ def hasDamIds(self, languageRecord, typeCode):
+
+		if language_record.Has_Text_Agreement() and type_code == "text":
+			self.add_licensor_if_not_empty(licensors, language_record.Licensor(), fileset_id, "licensor")
+			self.add_licensor_if_not_empty(licensors, language_record.CoLicensor(), fileset_id, "co-licensor")
```
@bradflood , it has been updated and I have removed the method add_licensor_if_not empty. You can see it the following commit: https://github.com/faithcomesbyhearing/dbp-etl/pull/130/commits/07f8396dc8bc1777b2f0a7d92b3aeeb4f7089aa2#diff-0628aedffa821d1b9acf4c81d50c756fe72674fcc7ca34c47e0a5b0a65c60aceR126

vichugofsl commented:
@bradflood , I have updated the PR according to your feedback and you can see the changes here: https://github.com/faithcomesbyhearing/dbp-etl/pull/130/commits/07f8396dc8bc1777b2f0a7d92b3aeeb4f7089aa2#diff-0628aedffa821d1b9acf4c81d50c756fe72674fcc7ca34c47e0a5b0a65c60aceR126

Also, I have updated the outcome in the PR description.
